Dockerfile

```diff
@@ -1,82 +1,164 @@
+        https://github.com/espressif/clang-xtensa.git "${LLVM_PATH}/tools/clang" \
+ && cd "${LLVM_PATH}/tools/clang/" \
+ && git reset --hard "${CLANG_VERSION}" \
+ && cd "${LLVM_PATH}" \
+ && git reset --hard "${LLVM_VERSION}" \
+ && mkdir -p "${LLVM_BUILD_PATH}" \
+ && cd "${LLVM_BUILD_PATH}" \
+ && cmake "${LLVM_PATH}" \
+       -DLLVM_TARGETS_TO_BUILD="Xtensa;X86" \
+       -DLLVM_INSTALL_UTILS=ON \
+       -DLLVM_BUILD_TESTS=0 \
+       -DLLVM_INCLUDE_TESTS=0 \
+       -DCMAKE_BUILD_TYPE=Release \
+       -DCMAKE_INSTALL_PREFIX="${LLVM_BASE}/llvm_install" \
+       -G "Ninja" \
+ && ninja install \
+ && rm -rf "${LLVM_PATH}" "${LLVM_BUILD_PATH}"
+
+# -------------------------------------------------------------------
+# Build rust-xtensa
+# -------------------------------------------------------------------
+
+WORKDIR "${RUSTC_BASE}"
+RUN git clone \
+        --recursive --single-branch \
+        https://github.com/MabezDev/rust-xtensa.git \
+        "${RUSTC_PATH}" \
+ && mkdir -p "${RUSTC_BUILD_PATH}" \
+ && cd "${RUSTC_PATH}" \
+ && git reset --hard "${RUSTC_VERSION}" \
+ && ./configure \
+        --llvm-root "${LLVM_INSTALL_PATH}" \
+        --prefix "${RUSTC_BUILD_PATH}" \
  && python ./x.py build \
- && python ./x.py install \
- && cd ..
-
-RUN /rust_build/bin/rustc --print target-list | grep xtensa
-RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
-ENV PATH=$PATH:/root/.cargo/bin
+ && python ./x.py install
 
-RUN rustup component add rustfmt
+# -------------------------------------------------------------------
+# Setup rustup toolchain
+# -------------------------------------------------------------------
 
-# install bindgen
-RUN cargo install bindgen
-# set LIBCLANG_PATH for bindgen
-ENV LIBCLANG_PATH=/llvm_build/lib
+RUN curl \
+        --proto '=https' \
+        --tlsv1.2 \
+        -sSf \
+        https://sh.rustup.rs \
+    | sh -s -- -y --default-toolchain stable \
+ && rustup component add rustfmt \
+ && rustup toolchain link xtensa "${RUSTC_BUILD_PATH}" \
+ && cargo install cargo-xbuild bindgen
 
-RUN rustup toolchain link xtensa /rust_build
-RUN rustup run xtensa rustc --print target-list | grep xtensa
+# -------------------------------------------------------------------
+# Our Project
+# -------------------------------------------------------------------
 
-# add xargo
-RUN cargo install xargo
-ENV XARGO_RUST_SRC=/rust-xtensa/src
+ENV PROJECT="/home/project/"
 
-# set up the build directory
-RUN mkdir /build
-VOLUME /build
-WORKDIR /build
+ENV XARGO_RUST_SRC="${RUSTC_PATH}/src"
+ENV TEMPLATES="${TOOLCHAIN}/templates"
+ENV LIBCLANG_PATH="${LLVM_INSTALL_PATH}/lib"
 
-COPY create-project image-project bindgen-project build-project xargo-project /usr/local/bin/
-RUN chmod a+x /usr/local/bin/*
-COPY templates /templates
+VOLUME "${PROJECT}"
+WORKDIR "${PROJECT}"
 
-CMD /usr/local/bin/build-project
+COPY bindgen-project build-project create-project image-project xbuild-project /usr/local/bin/
```

Comment:
I just realized that this PR didn't include the `flash-project` script into this copy command. You may want to add it if the current running build succeed.

Reply:
Can you amend the PR?

Reply:
Done